### PR TITLE
[BC-breaking] Change default eps value of FrozenBN

### LINKED ISSUE #2599: Change default value of eps in FrozenBatchNorm to match BatchNorm
## ❓ Questions and Help
Hello
Loss is nan error occurs when I learn fast rcnn with resnext101 backbone
My code is as follows
```python
backbone = resnet_fpn_backbone('resnext101_32x8d', pretrained=True)
model = FasterRCNN(backbone, num_classes)
in_features = model.roi_heads.box_predictor.cls_score.in_features
model.roi_heads.box_predictor = FastRCNNPredictor(in_features, num_classes)
```

error message
```
Epoch: [0]  [   0/7208]  eta: 1:27:42  lr: 0.000040  loss: 40613806080.0000 (40613806080.0000)  loss_box_reg: 7979147264.0000 (7979147264.0000)  loss_classifier: 11993160704.0000 (11993160704.0000)  loss_objectness: 9486380032.0000 (9486380032.0000)  loss_rpn_box_reg: 11155118080.0000 (11155118080.0000)  time: 0.7301  data: 0.4106  max mem: 1241
Loss is nan, stopping training
```

When i change the backbone to resnet50 and resnet152, no error occrus.
### Please note that this issue tracker is not a help form and this issue will be closed.

We have a set of [listed resources available on the website](https://pytorch.org/resources). Our primary means of support is our discussion forum:

- [Discussion Forum](https://discuss.pytorch.org/)

### PR DESCRIPTION
**Pull request by datumbox**
Fixes #2599

The good news is that changing the `FrozenBatchNorm2d` default `eps` value from `0.0` to `1e-5` does not affect the accuracy of the detection models. You will see some tiny variations from the [reported metrics](https://github.com/pytorch/vision/blob/master/docs/source/models.rst) but nothing concerning (could be the result of floating point errors). So I believe we can safely land the change without problems.

<details><summary>Click here for the accuracy stats of fasterrcnn_resnet50_fpn</summary>

```
IoU metric: bbox
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.369
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.585
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.398
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.210
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.403
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.480
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.307
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.485
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.509
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.315
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.545
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.644
```
</details>

<details><summary>Click here for the accuracy stats of retinanet_resnet50_fpn</summary>

```
IoU metric: bbox
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.363
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.557
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.382
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.191
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.400
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.489
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.314
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.500
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.539
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.339
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.580
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.695
```
</details>

<details><summary>Click here for the accuracy stats of maskrcnn_resnet50_fpn</summary>

```
IoU metric: bbox
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.379
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.591
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.410
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.215
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.414
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.494
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.313
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.495
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.519
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.321
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.559
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.666
IoU metric: segm
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.346
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.560
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.368
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.156
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.373
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.507
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.295
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.455
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.475
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.269
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.515
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.636
```
</details>

<details><summary>Click here for the accuracy stats of keypointrcnn_resnet50_fpn</summary>

```
IoU metric: bbox
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.546
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.830
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.597
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.380
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.628
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.702
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.187
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.555
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.636
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.493
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.700
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.775
IoU metric: keypoints
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.651
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.861
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.716
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.604
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 0.730
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 20 ] = 0.717
 Average Recall     (AR) @[ IoU=0.50      | area=   all | maxDets= 20 ] = 0.906
 Average Recall     (AR) @[ IoU=0.75      | area=   all | maxDets= 20 ] = 0.774
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets= 20 ] = 0.670
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets= 20 ] = 0.784
```
</details>

The bad news is that the following detection unit-tests are brittle and changing the `eps` value causes them to break:
- test_keypointrcnn_resnet50_fpn_cpu
- test_keypointrcnn_resnet50_fpn_cuda
- test_retinanet_resnet50_fpn_cpu

I tried updating the expecting values to new ones but unfortunately there is too much variation across different devices and platforms. As observed at #2812, the flakiness is caused by floating point errors and the use of unstable sort. The problem is the result of initializing the model weights with random values and using random input in the unit-tests. 

To fix the problem I introduce a VERY [hacky workaround](https://github.com/pytorch/vision/pull/2933/commits/7c8c5ae90653edd17997b9451a01db757f93df00) where I switch the default value to 0.0 for specific unit-tests. I really don't like this hack but I don't see how else we can keep the unit-tests on and while changing the `eps` . If anyone has another idea, I'm happy to give it a try.

### CODE REVIEW

File: test/test_models.py [hunk 1]
```diff
@@ -148,7 +148,20 @@ def _test_detection_model(self, name, dev):
...
+        class OverwriteEPS:
+            def __enter__(self):
+                self.default = FrozenBatchNorm2d._DEFAULT_EPS
+                FrozenBatchNorm2d._DEFAULT_EPS = 0.0
+
+            def __exit__(self, type, value, traceback):  # noqa: F811
+                FrozenBatchNorm2d._DEFAULT_EPS = self.default
+
+        with OverwriteEPS():
+            model = models.detection.__dict__[name](num_classes=50, pretrained_backbone=False, **kwargs)
```
Comment (fmassa) [inline test/test_models.py:164]: Instead of doing this workaround (and having to introduce test-specific code in the op implementation, what about replacing all occurrences of `FrozenBatchNorm.eps` with the value 0?

Something like
```python
model = models.detection.__dict__[name](num_classes=50, pretrained_backbone=False, **kwargs)
for module in model.modules():
    if isinstance(module, FrozenBatchNorm2d):
        module.eps = 0
```

**Review by fmassa — CHANGES_REQUESTED** [review]
Thanks for the PR!

It's great that the changes on accuracy don't seem to be affected in a significant way.
I suspect the results before the change are also the same values, or does it decrease mAP by 0.1?

I have one remark about the hack, I think it can be made such that we don't need to change the implementation in the op itself

File: test/test_models.py (hunk 1 above)
Comment (datumbox) [reply test/test_models.py]: I like this better than I like mine workaround. Still we are putting lipstick on a pig but it will work for now. :)

**Comment by datumbox** [comment]
Here are the baseline vs new comparisons:

**fasterrcnn_resnet50_fpn:**
- MASTER: boxAP=0.368
- BRANCH: boxAP=0.369

**retinanet_resnet50_fpn:**
- MASTER: boxAP=0.364 
- BRANCH: boxAP=0.363

**maskrcnn_resnet50_fpn:**
- MASTER: boxAP=0.378 / maskAP=0.344
- BRANCH: boxAP=0.379 / maskAP=0.346

**keypointrcnn_resnet50_fpn:**
- MASTER: boxAP=0.546 / kpAP=0.650
- BRANCH: boxAP=0.546 / kpAP=0.651

So overall I think we are OK.

**Review by fmassa — APPROVED** [review]
Awesome, thanks a lot!
